presto-orc/src/main/java/com/facebook/presto/orc/OrcRecordReader.java

```
@@ -761,27 +770,89 @@ public static LinearProbeRangeFinder createTinyStripesRangeFinder(List<StripeInf
+            }
+        }
+        nonDeterministicConstantFilters = nonDeterministicBuilder.build();
         reader = new ColumnGroupReader(
                 streamReaders,
                 presentColumns,
                 channelColumns,
                 types,
                 options.getInternalChannels(),
                 options.getOutputChannels(),
-                predicate.getFilters(),
-                options.getFilterFunctions(),
+                filters,
+                anyConstantFilterFunctions ? Arrays.stream(filterFunctions).filter(f -> !isConstantFilterFunction(f, constantBlocks)).toArray(FilterFunction[]::new) : filterFunctions,
                 enforceMemoryBudget,
-                reorderFilters);
+                constantBlocks);
         targetResultBytes = options.getTargetBytes();
         reader.setResultSizeBudget(targetResultBytes);
     }
 
+    private static boolean isConstantFilterFunction(FilterFunction filter, Block[] constantBlocks)
+    {
+        return Arrays.stream(filter.getInputChannels()).allMatch(channel -> constantBlocks[channel] != null);
+    }
+
+    private static void evaluateConstantFilterFunction(FilterFunction filter, Block[] constantBlocks, QualifyingSet qualifyingSet)
+    {
+        int[] channels = filter.getInputChannels();
+        Block[] inputs = new Block[channels.length];
+        for (int i = 0; i < channels.length; i++) {
+            inputs[i] = constantBlocks[channels[i]];
+        }
+        int[] filterResults = new int[qualifyingSet.getPositionCount()];
+        PageSourceOptions.ErrorSet errors = qualifyingSet.getOrCreateErrorSet();
+        int numHits = filter.filter(new Page(qualifyingSet.getPositionCount(), inputs), filterResults, errors);
+        qualifyingSet.compactPositionsAndErrors(filterResults, numHits);
+    }
+
     public Page getNextPage()
             throws IOException
     {
+        if (constantFilterIsFalse) {
+            return null;
```

I believe there's an invariant that `pushdownFilterAndProjection` is always called before `getNextPage`.  If that's true can we throw an exception if that is violated?